### DISCUSSION TITLE
Proposed revision of mods introduced in Lazer

### DISCUSSION THREAD

**Discussion by mk56-spn** [ppy/osu#18905] [Ideas]
After extensive play of all new mods in Lazer, and most of the new options added to old mods, I'm of the opinion that it would be useful to list all the issues in them more related to enjoyability.

the quality control of the mods up til now has been mostly focused on code quality, which is of course important, but i feel the gameplay quality isn't there in many of them, the issues vary from merely irksome in nature to hampering the playing experience.

Here's a list of the issues, i'm hoping for second opinions from other Lazer players of these to ascertain whether its personal bias or legitimate.

<details open><summary> 

#### Deflate

 </summary>
 The maximum size seems too big, probably should be toned down from 25x to 10x 
<img src="https://user-images.githubusercontent.com/74463310/175971921-3709ef53-202a-4130-ab16-ca6fdf6ba640.jpg">
</details>
<details open><summary> 

#### Traceable

 </summary>
Currently slider heads are way easier to acc and hit since you can use the edge of the slider body as a guide, which makes the difference between sliderheads and circles feel janky.
<img src="https://user-images.githubusercontent.com/74463310/175972370-46048fd4-ef6e-494a-bf6f-895b6fd0b464.jpg">
#13137 this old discussion had a very nifty solution which fits the theme aesthetic wise and keeps the sliders balanced

Remove followpoints from traceable or give traceable custom follow points with limited length to avoid cheese skins such as this:

<img src="https://user-images.githubusercontent.com/74463310/181855682-641dcaa0-76ce-4a4c-a29c-0dd107a61fcb.png">
</details>
<details open><summary> 

Make a custom enforced approach circle for use in traceable to avoid people using  approach circles such as this: 

https://user-images.githubusercontent.com/74463310/213166394-253614c6-13e0-4f80-825e-443057779474.mp4

since it kills any semblance of uniqueness in the mod. making it just deflate II

#### Approach different

</summary>

Mostly fine but could be helped with better names for the menu, i struggle to remember what "accelerate 1" does differently from "accelerate 2", as such having the proper names of the easing factors (as in "in out quint" instead of "in out 2") would help.<br />
<img src="https://user-images.githubusercontent.com/74463310/175973423-c7a8622a-11da-4de7-86fe-229241e8bdcc.png">

</details>
<details open><summary> 

#### Random Mod

</summary>

- Sliders should be able to flip themselves not just rotate

- Sometimes random makes slider ends unreasonably harder to hit by pointing sliders ends that were initially pointed at the next circle/ slider point off into nowhere, limiting the deviation angle in these cases would make the maps a lot more fun, especially on maps with long sliders.

- Random would benefit greatly in readability from not allowing sliders that previously did not overlap to not be able to do so after random is applied

- Could do with an option to auto randomise seeds between each map restart, possibly interesting from the point of view of tourneys as well?

</details>
<details open><summary> 

#### Spin in

</summary>
No particular issues with the functionality but it is in my opinion not very enjoyable, and could do with an overhaul, something akin to a clock arrow spinning inside the circle a full rotation before the player is meant to press it
</details>
<details open><summary> 

#### Mirror

</summary> 
As with random an option to scramble its settings would be useful, since i assume this will mostly be used to break mind blocks, hence having an option to have it randomise its settings is nice

</details>
<details><summary> 

#### New Flashlight settings (Done)

</summary>

- The softness of the edge varies with the scaling of the fl radius, i don't know if this is something flashlight players would enjoy, dislike, or be indifferent to.
<img src="https://user-images.githubusercontent.com/74463310/175974806-580c27b5-966b-402d-a359-37e674cc71ad.jpg">
<img src="https://user-images.githubusercontent.com/74463310/175974868-00e05d8c-d2ed-40ab-b163-fa06763ed6c3.jpg">

</details>

<details open><summary> 

#### Difficulty adjust

</summary>

- Currently the mod wont set a custom value if its value overlaps with with the maps default value, in my experience this runs counter to my most common usage, which is setting it  a value for a  multitude of maps, its a minor inconvenience at most but being able to set ar9 without switching off of an ar9 map would be nice

- Adding stack leniency to difficulty adjust
</details>
<details open><summary> 

#### Blinds

</summary>

- Possibly untying blinds from HP, in favour of something like note density or accuracy, given HP is rather unreliable as a metric for anything and makes the movements of the blinds feel very janky.

- Horizontal option would be nice to choose depending on map type.
</details>
<details open><summary> 

#### Alternate mod

</summary>
Alternate could do with an option to have the notes alternate between two colours, seeing as its a learning oriented mod, i think this would ease the experience for many
</details>
<details open><summary> 

#### Hidden

</summary>
The fadeoutless option is currently quite awkward, #15220 would help cover this, it's hardly mandatory, but would improve the playability and fairness of the mod, that or allowing metronome sounds similarly to muted
</details>
<details open><summary> 

#### Traceable approach circle fade option ( https://github.com/ppy/osu/discussions/18114 PRed : https://github.com/ppy/osu/pull/19845)

</summary>
<video src="https://user-images.githubusercontent.com/74463310/181801077-37f4c942-d516-446f-8262-d6f940634574.mp4">
</details>
<details open><summary> 

#### Wind up/down

</summary>
Add ability to lock AR and OD

<details open><summary> 

#### Followpoint issues ( **Pred by  @o-dasher , https://github.com/ppy/osu/pull/20514**)

</summary>

#### Freeze frame

</summary>
find a way to allow compatibility with approach different 

<details open><summary> 

#### Followpoint issues ( **Pred by  @o-dasher , https://github.com/ppy/osu/pull/20514**)

</summary>

- Maybe add a subsetting for some mod ( DA? ) that forces follow points off? this allows for players to choose whether they want a harder but more rewarding experience( iirc followpoints are considered for pp ) or an easier one.
<img src="https://user-images.githubusercontent.com/74463310/183478062-a5d11d8d-c486-49d3-9e96-cec6be656dbf.jpg">

<img src="https://user-images.githubusercontent.com/74463310/183478087-18c01e59-5f3f-461a-9ec3-b820d224f10e.jpg">

(Here's an example to demonstrate just how big a difference it makes, in example  a you can clearly see the vector you're meant to be aiming along, leaving you to just learn the distance, in the second example you basically have to memorise the position for the notes ( except for the few that come into view long enough before you have to hit them)

</details>

**Comment by mk56-spn** [comment]


**Reply by frenzibyte** [reply]
Please update description rather than posting an "update", this causes unnecessary noise.

**Reply by mk56-spn** [reply]
> Please update description rather than posting an "update", this causes unnecessary noise.

Oops. My bad. Thought it didn't matter since discussions don't get shown on discord

**Reply by frenzibyte** [reply]
It piles up on our GitHub notifications list.

**Reply by peppy** [reply]
It also helps keep things in one place. For instance, when I (or someone else) eventually gets to going through this list, having all of the items in a single post makes it easier to parse or copy elsewhere.

Please make sure to edit your original post with the changes you've made in this reply.

**Reply by mk56-spn** [reply]
Will do !

**Comment by ggliv** [comment]
Big 'ol discussion post, just going to chime in with some of my thoughts where I have them to hopefully help get a consensus going for potential changes.

Deflate: Your suggestion seems reasonable to me. Most of what you'd get from having a starting size of 25 is preserved if you set it to 10.

Traceable: I personally think the current inclusion of slider bodies is fine. Masking out slider elements within the approach circle adds (in my view) unnecessary complexity and could significantly affect the readability of shorter sliders that would only be shown as the approach circle closes in on them. Might be wrong though, I'd have to give it a play first.

Approach Different: I feel like the animation variations (InOut1, InOut2) could be removed in favor of choosing one to represent the whole bunch (just InOut). Alternatively, they could be renamed to [reflect the easing functions they end up calling](https://github.com/ppy/osu/blob/f15698d02535c3c8be7f2453c5dcffd0e9dda71b/osu.Game.Rulesets.Osu/Mods/OsuModApproachDifferent.cs#L57-L85) (i.e. InOut2 -> InOut Quint).

Random/Mirror: Setting randomization could probably:tm: be handled through an interface and applied as appropriate, I could definitely see this being useful/fun for mods like Approach Different or even Difficulty Adjust.

Spin In: I have a similarly lukewarm view of this mod. Definitely would not mind a rework, but I have no strong opinions on how something like that would go. Your clock proposal seems sane.

Alternate: Any implementation of this color-coding would have to account for cases where a player misses a hitobject and is then forced to continue using the same button for the next one, likely by dynamically changing the color pattern. Because this may be jarring to some users and a number of beatmaps have very intentionally selected hitobject color schemes, I would suggest that something like this be a toggle that is defaulted to off.

Hidden: @mk56-spn Just to clarify, by "the proposed snap colouration settings" are you referring to https://github.com/ppy/osu/discussions/15220?

Hidden-Traceable compatibility: I don't love the idea of changing a mod's behavior for just one pairing, IMO it's fine to keep them incompatible.

**Reply by mk56-spn** [reply]
i agree with with most of these and ill update my post to reflect such in a bit

i wont bother commenting on the things youve said i agree with , since it would add nothing

i found an old post that got forgotten which had a better solution for the traceable slider issue, i dont feel how it is right now is a reasonable compromise but that one is. basically it reduces sliders to a line at the center of their radius, which preserves readability, which as you pointed out my old method didnt, whilst obscuring the " when to click factor  https://github.com/ppy/osu/discussions/13137 here it is more in depth 

hidden traceable compatibility : how about a simple "fade approach circles " under traceables settings? , whilst making the mods incompatible, would that be a compromise you are willing to accept? 

Alternate : hmm needs more thought, i agree with the issue you pointed out, if it comes it should def be  optional, this was more of just a helper for new players than something id expect everyone to use, i full alt regardless

Dk about randomisation through another interface, i can see the pros but implementation might be taxing unless you already have a plan for it?.

Hidden: i am indeed referring to #15220 and its many offshoot issue and discussions

I agree with naming the approach different settings as mentioned with the  easing functions, some are pretty different and removing them feels a bit harsh IMO, but im not totally staunch on this issue

sidenote: thanks for the  useful feedback, its nice to have more eyes on the matter!

**Reply by ggliv** [reply]
Adding a 'slim' slider for Traceable seems like it should be a good solution. I have no immediate oppositions to adding a fading approach circle option either.

I figure we should start with trying to rename the AD animation styles and only proceed to removing some of them if people open further issues or discussions confused about their differences.

I'm currently away from my normal workstation and will be for several more days but I'll try to open PRs for some of these as I get the chance.

**Reply by mk56-spn** [reply]
> Adding a 'slim' slider for Traceable seems like it should be a good solution. I have no immediate oppositions to adding a fading approach circle option either.
> 
> I figure we should start with trying to rename the AD animation styles and only proceed to removing some of them if people open further issues or discussions confused about their differences.
> 
> I'm currently away from my normal workstation and will be for several more days but I'll try to open PRs for some of these as I get the chance.

Thanks for the help, I'm a rather mediocre code monkey to say the least, I can handle the renaming though so I'll leave the hard stuff to you
